src/test/client/nonblocking.cc

```diff
@@ -611,3 +612,589 @@ TEST_F(TestClient, LlreadvLlwritevInvalidFileHandle) {
+                         iov_out_dataonly_fysnc[i].iov_len));
+  }
+
+  client->ll_release(fh);
+  ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));
+}
+
+TEST_F(TestClient, LlreadvLlwritevBufferOverflow) {
+  /* Provide empty read buffers to see how the function behaves
+  when there is no space to store the data*/
+
+  int mypid = getpid();
+  char filename[256];
+
+  client->unmount();
+  TearDown();
+  SetUp();
+
+  sprintf(filename, "test_llreadvllwritevbufferoverflowfile%u", mypid);
+
+  Inode *root, *file;
+  root = client->get_root();
+  ASSERT_NE(root, (Inode *)NULL);
+
+  Fh *fh;
+  struct ceph_statx stx;
+
+  ASSERT_EQ(0, client->ll_createx(root, filename, 0666,
+				  O_RDWR | O_CREAT | O_TRUNC,
+				  &file, &fh, &stx, 0, 0, myperm));
+
+  char out_buf_0[] = "hello ";
+  char out_buf_1[] = "world\n";
+  struct iovec iov_out[2] = {
+	  {out_buf_0, sizeof(out_buf_0)},
+	  {out_buf_1, sizeof(out_buf_1)},
+  };
+
+  char in_buf_0[0];
+  char in_buf_1[0];
+  struct iovec iov_in[2] = {
+	  {in_buf_0, sizeof(in_buf_0)},
+	  {in_buf_1, sizeof(in_buf_1)},
+  };
+
+  ssize_t bytes_to_write = iov_out[0].iov_len + iov_out[1].iov_len;
+
+  std::unique_ptr<C_SaferCond> writefinish = nullptr;
+  std::unique_ptr<C_SaferCond> readfinish = nullptr;
+
+  writefinish.reset(new C_SaferCond("test-nonblocking-writefinish-buffer-overflow"));
+  readfinish.reset(new C_SaferCond("test-nonblocking-readfinish-buffer-overflow"));
+
+  int64_t rc;
+  bufferlist bl;
+
+  rc = client->ll_preadv_pwritev(fh, iov_out, 2, 0, true, writefinish.get(),
+                                 nullptr);
+  ASSERT_EQ(rc, 0);
+  ssize_t bytes_written = writefinish->wait();
+  ASSERT_EQ(bytes_written, bytes_to_write);
+
+  rc = client->ll_preadv_pwritev(fh, iov_in, 2, 0, false, readfinish.get(),
+                                 &bl);
+  ASSERT_EQ(rc, 0);
+  ssize_t bytes_read = readfinish->wait();
+  ASSERT_EQ(bytes_read, 0);
+  ASSERT_EQ(bl.length(), 0);
+
+  client->ll_release(fh);
+  ASSERT_EQ(0, client->ll_unlink(root, filename, myperm));
+}
+
+TEST_F(TestClient, LlreadvLlwritevQuotaFull) {
+  /*Test that if the max_bytes quota is exceeded, async I/O code path handles
+  the case gracefully*/
+
+  client->unmount();
+  TearDown();
+  SetUp();
+
+  Inode *root, *diri;
+  struct ceph_statx stx;
+  root = client->get_root();
+  ASSERT_NE(root, nullptr);
+
+  int pid = getpid();
+
+  char dirname[256];
+  sprintf(dirname, "testdirquotaufull%u", pid);
+  ASSERT_EQ(0, client->ll_mkdirx(root, dirname, 0777, &diri, &stx, 0, 0,
+                                 myperm));
+
+  // set quota.max_bytes
+  char xattrk[128];
+  sprintf(xattrk, "ceph.quota.max_bytes");
+  char xattrv[128];
+  int32_t len = sprintf(xattrv, "8388608"); // 8MiB
+  int64_t rc = client->ll_setxattr(diri, xattrk, xattrv, len,
+                                   CEPH_XATTR_CREATE, myperm);
+  ASSERT_EQ(rc, 0);
+  rc = client->ll_getxattr(diri, xattrk, (void *)xattrv, len, myperm);
+  ASSERT_EQ(rc, 7);
```

Comment:
If the value has to be checked, validate the value is what was set earlier.

Reply:
uh sorry. let me add it

Reply:
done